covalent/executor/base.py

```diff
@@ -467,3 +462,48 @@ def get_conda_path(self) -> bool:
             return False
         self.conda_path = which_conda
         return True
+
+
+class BaseAsyncExecutor(BaseExecutor):
+    async def execute(
+        self,
+        function: Callable,
+        args: List,
+        kwargs: Dict,
+        dispatch_id: str,
+        results_dir: str,
+        node_id: int = -1,
+    ) -> Any:
+        awaitable_run, out, err = super().execute(
+            function, args, kwargs, dispatch_id, results_dir, node_id
+        )
+
+        if not asyncio.iscoroutine(awaitable_run):
+            return (awaitable_run, out, err)
```

Comment:
What do you think of making  `BaseAsyncExecutor` its own base class instead of a subclass of `BaseExecutor`? That way we can enforce that `run()` is a coroutine. Trying to handle the case where `run()` is not a coroutine here would block the event loop.

Reply:
We can make a simple `BaseAsyncExecutor` base class with just `execute` and `run`. All the logic pertaining to a local conda env can be jettisoned as it is incompatible with async.

Reply:
Since an async `run()` is not really "derived" from non-async `run()` (they aren't invoked in the same way), `BaseAsyncExecutor` is logically parallel to, not derived from, `BaseExecutor`.

Reply:
>What do you think of making `BaseAsyncExecutor` its own base class instead of a subclass of `BaseExecutor`?

I would've agreed with that, actually I was going to do it like that at first, but a lot of things in `BaseAsyncExecutor` are the same as in `BaseExecutor` which are unrelated to `conda`, and their respective function signature, attribute types are similar as well; things like: `log_stdout`, `log_stderr`, `cache_dir`, `get_dispatch_context` (although this will be removed soon), `write_streams_to_file()`, `to_dict()`, `from_dict()`, `short_name()`.

Since all of these functions/attributes will need to be copy-pasted in the new `BaseAsyncExecutor`, I figured it was better to derive them from `BaseExecutor`. Perhaps if we put those things in a common file then both executors can import it from there? Might be out of the scope of this issue.

>That way we can enforce that run() is a coroutine. Trying to handle the case where run() is not a coroutine here would block the event loop.

Yeah, I agree that it will block the eventloop here, but only in the case when the `conda_env` is specified.
Will creating a separate class allow us to enforce `run()` being a coroutine in a way which cannot be done now?

>Since an async `run()` is not really "derived" from non-async `run()` (they aren't invoked in the same way), `BaseAsyncExecutor` is logically parallel to, not derived from, `BaseExecutor`.

Agreed, but it's actually the other stuff that is derived from `BaseExecutor` not `run()`.

Reply:
I also like the idea of replacing how conda related stuff is being handled right now in executors. I think we can improve it.

Reply:
>Since all of these functions/attributes will need to be copy-pasted in the new BaseAsyncExecutor, I figured it was better to derive them from BaseExecutor. Perhaps if we put those things in a common file then both executors can import it from there? Might be out of the scope of this issue.

Reply:
This restructuring can be left for a followup issue. I'm happy to approve this for now.

Reply:
>Maybe something like
`AbstractBaseExecutor` (common methods)
`BaseExecutor` `BaseAsyncExecutor`

Yeah, I like that.